### PR TITLE
Adding refresh_token to GoogleOAuthenticator

### PR DESCRIPTION
**Pull request by missingcharacter**
I've been working on giving my users a unified experience using G Suite, we ran into a situation where we wanted to have access to files in Google Drive and most applications that interact with Google Drive (Example: https://github.com/astrada/google-drive-ocamlfuse) require a `refresh_token` in order to work.

Since users already authenticate to Google when they login to jupyterhub I thought it was best to do all at once. 

These are the changes and I made and I thought the rest of the community would benefit from them too.

### CODE REVIEW

**Comment by missingcharacter** [comment]
Looks like the error is not related to my changes:

```
oauthenticator/github.py:154:34: F523 '...'.format(...) has unused arguments at position(s): 0
The command "if [[ "$TEST_LINT" = 1 ]]; then
  flake8 oauthenticator
else
  py.test --cov oauthenticator oauthenticator
fi
" exited with 1.
```

I can try to take a look at what happen, but I did not touch `github.py`

**Comment by betatim** [comment]
Hey, thanks for the PR.

Could you write a short summary of the changes, in particular "why" making them is a good idea and/or what problem they solve in the top comment of this PR? This will help reviewers right now as well as people coming here via search engines or the future who are trying to understand what the heck was done and why.

File: oauthenticator/google.py [hunk 1]
```diff
@@ -27,6 +28,16 @@ def check_user_in_groups(member_groups, allowed_groups):
...
         return False
 
+class GoogleLoginHandler(OAuthLoginHandler):
+    """See https://developers.google.com/identity/protocols/oauth2 for general information."""
+
+    def authorize_redirect(self, *args, **kwargs):
+        """Add `access_type`, `approval_prompt` to redirect params"""
+        extra_params = kwargs.setdefault('extra_params', {})
+        extra_params["access_type"] = 'offline'
+        extra_params["approval_prompt"] = 'force'
```
Comment (betatim) [inline oauthenticator/google.py:38]: can you explain a bit why this change is needed and what it will do? Especially what effect will it have on people already using this authenticator.

At first glance it seems unrelated to storing the refresh token.

File: oauthenticator/google.py [hunk 2]
```diff
@@ -167,9 +181,30 @@ async def authenticate(self, handler, data=None, google_groups=None):
                 # unambiguous domain, use only base name
                 username = user_email.split('@')[0]
 
+        if refresh_token == None:
```
Comment (betatim) [inline oauthenticator/google.py:184]:
```suggestion
        if refresh_token is None:
```

Comment (missingcharacter) [reply oauthenticator/google.py]: done with this

File: oauthenticator/google.py (hunk 1 above)
Comment (missingcharacter) [reply oauthenticator/google.py:38]: These changes are to get an `access_token` and `refresh_token` for all scopes at once. In the case that people are adding more scopes to their login request, example: to give users access to Google Drive.

**Comment by missingcharacter** [comment]
@betatim Sorry, I did not make the case for `why?` before, I added it now.

File: oauthenticator/google.py (hunk 1 above)
Comment (missingcharacter) [reply oauthenticator/google.py:38]: I could also work on making these something people opt-in

Comment (missingcharacter) [reply oauthenticator/google.py:38]: maybe something like this to make it something people opt-in

```diff
diff --git a/oauthenticator/google.py b/oauthenticator/google.py
index b5b1e31..7fc3bf3 100644
--- a/oauthenticator/google.py
+++ b/oauthenticator/google.py
@@ -13,7 +13,7 @@ from tornado.httpclient import HTTPRequest, AsyncHTTPClient
 from tornado.auth import GoogleOAuth2Mixin
 from tornado.web import HTTPError

-from traitlets import Dict, Unicode, List, default, validate
+from traitlets import Bool, Dict, Unicode, List, default, validate

 from jupyterhub.crypto import decrypt, EncryptionUnavailable, InvalidToken
 from jupyterhub.auth import LocalAuthenticator
@@ -31,11 +31,17 @@ def check_user_in_groups(member_groups, allowed_groups):
 class GoogleLoginHandler(OAuthLoginHandler):
     """See https://developers.google.com/identity/protocols/oauth2 for general information."""

+    retrieve_all_scopes_at_once = Bool(
+        os.environ.get('GOOGLE_LOGIN_HANDLER_ALL_SCOPES', 'False').lower() in {'true', '1'},
+        help="Set extra_params `access_type` to `offline` and `approval_prompt` to `force`"
+    ).tag(config=True)
+
     def authorize_redirect(self, *args, **kwargs):
         """Add `access_type`, `approval_prompt` to redirect params"""
-        extra_params = kwargs.setdefault('extra_params', {})
-        extra_params["access_type"] = 'offline'
-        extra_params["approval_prompt"] = 'force'
+        if self.retrieve_all_scopes_at_once:
+            extra_params = kwargs.setdefault('extra_params', {})
+            extra_params["access_type"] = 'offline'
+            extra_params["approval_prompt"] = 'force'

         return super().authorize_redirect(*args, **kwargs)
```

Comment (betatim) [reply oauthenticator/google.py:38]: I don't have a good overview of how people currently/typically configure the Google authenticator. But opt-in sounds like a good way to reduce the chances of breaking stuff for people who are currently happily running the authenticator

Comment (manics) [reply oauthenticator/google.py:38]: I guess it's a balance between configurability and code complexity.

Is it easy for you to post some screenshots of the behaviour with and without this PR?

Comment (missingcharacter) [reply oauthenticator/google.py]: @manics the screenshots are the same with and without, the difference is that without these I do not get a refresh token and with them I do

Comment (missingcharacter) [reply oauthenticator/google.py]: logs with:
```
[D 2020-05-05 08:28:29.910 JupyterHub google:41] retrieve_all_scopes_at_once is set to True
[I 2020-05-05 08:28:29.911 JupyterHub log:174] 302 GET /hub/oauth_login?next= -> https://accounts.google.com/o/oauth2/v2/auth?response_type=code&redirect_uri=https%3A%2F%2FREDACTED%3A8000%2Fhub%2Foauth_callback&client_id=REDACTED&state=[secret]&access_type=offline&approval_prompt=force&scope=openid+email+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive (@::ffff:127.0.0.1) 1.58ms
[D 2020-05-05 08:30:18.805 JupyterHub google:157] user contains {'access_token': 'REDACTED', 'expires_in': 3599, 'refresh_token': 'REDACTED', 'scope': 'https://www.googleapis.com/auth/userinfo.email openid https://www.googleapis.com/auth/drive', 'token_type': 'Bearer', 'id_token': 'REDACTED'}
[D 2020-05-05 08:30:20.057 JupyterHub google:222] user_info contains {'name': 'person', 'auth_state': {'access_token': 'REDACTED', 'refresh_token': 'REDACTED', 'google_user': {'id': 'REDACTED', 'email': 'REDACTED', 'verified_email': True, 'picture': 'REDACTED', 'hd': 'example.com', 'google_groups': ['agroup', 'anothergroup', 'andanotherone']}}, 'admin': True}
```

logs without

```
[D 2020-05-05 08:46:58.945 JupyterHub google:41] retrieve_all_scopes_at_once is set to True
[I 2020-05-05 08:46:58.947 JupyterHub log:174] 302 GET /hub/oauth_login?next= -> https://accounts.google.com/o/oauth2/v2/auth?response_type=code&redirect_uri=https%3A%2F%2FREDACTED%3A8000%2Fhub%2Foauth_callback&client_id=REDACTED&state=[secret]&scope=openid+email+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive (@::ffff:127.0.0.1) 2.13ms
[D 2020-05-05 08:47:16.677 JupyterHub google:157] user contains {'access_token': 'REDACTED', 'expires_in': 3599, 'scope': 'openid https://www.googleapis.com/auth/userinfo.email https://www.googleapis.com/auth/drive', 'token_type': 'Bearer', 'id_token': 'REDACTED'}
[D 2020-05-05 08:47:16.727 JupyterHub google:194] Refresh token was empty, will try to pull refresh_token from previous auth_state
[D 2020-05-05 08:47:16.734 JupyterHub google:198] encrypted_auth_state was found, will try to decrypt and pull refresh_token from it
[D 2020-05-05 08:47:17.626 JupyterHub google:222] user_info contains {'name': 'person', 'auth_state': {'access_token': 'REDACTED', 'refresh_token': 'REDACTED', 'google_user': {'id': 'REDACTED', 'email': 'REDACTED', 'verified_email': True, 'picture': 'REDACTED', 'hd': 'example.com', 'google_groups': ['agroup', 'anothergroup', 'andanotherone']}}, 'admin': True}
```

Comment (missingcharacter) [reply oauthenticator/google.py]: The most important part is `user contains` I added this line `self.log.debug("user contains %s", user)` right after this line https://github.com/jupyterhub/oauthenticator/blob/75a383e88073f1e58e1fd996a3271aae2c256f29/oauthenticator/google.py#L136

Comment (missingcharacter) [reply oauthenticator/google.py]: Forgot to add that I revoked access to Jupyterhub each time I tried the steps described above.

**Comment by manics** [comment]
Another option we could consider, are people likely to want to customise the values you've added in `extra_params`, or add other options? If so would it be more extensible to make `extra_params` a configuration property (Dictionary)?

**Comment by missingcharacter** [comment]
@manics thanks for the suggestion this is better

**Comment by NickolausDS** [comment]
It might be nice to include a link to the Google Docs to list the other possible values folks can add for `extra_params`. I think their main link is here: https://developers.google.com/identity/protocols/oauth2/web-server#creatingclient

Adding `extra_params` would be useful for other OAuth2 providers as well. [CILogin](https://github.com/jupyterhub/oauthenticator/blob/master/oauthenticator/cilogon.py#L32) also overrides the LoginHandler to add extra params. [Github](https://developer.github.com/apps/building-oauth-apps/authorizing-oauth-apps/#parameters)  allows `allow_signup`. [Globus](https://docs.globus.org/api/auth/reference/#authorization_code_grant_preferred) also supports `access_type=offline` for Refresh Tokens. #338 would allow a general configuration property for `extra_params` without overriding a LoginHandler.

**Comment by missingcharacter** [comment]
@NickolausDS I added the link you suggested, thanks.

**Comment by missingcharacter** [comment]
@manics @betatim are there any other changes I should make to this?

**Comment by manics** [comment]
If https://github.com/jupyterhub/oauthenticator/pull/338 is merged will that mean this PR can use the parent variable instead of defining it in this class?

**Comment by missingcharacter** [comment]
@manics correct, I would only leave the handling of the refresh token if found

**Comment by missingcharacter** [comment]
@manics I've rebased from `master` after https://github.com/jupyterhub/oauthenticator/pull/338 was merged

File: docs/source/google.md [hunk 3]
```diff
@@ -63,3 +63,13 @@ c.GoogleOAuthenticator.google_service_account_keys = {'example.com': '/path/to/s
 c.GoogleOAuthenticator.google_group_whitelist = {'example.com': ['somegroupwithaccess', 'othergroupwithaccess'] }
 ```
 ### You are done!
+
+## How to retrieve an `access_token` and `refresh_token` for all scopes at once
+
+In your `jupyterhub_config.py` do the following:
+
+```python
+c.JupyterHub.authenticator_class.extra_authorize_params = {'access_type': 'offline', 'approval_prompt': 'force'}
```
Comment (minrk) [inline docs/source/google.md:72]: one minor comment:

```suggestion
c.OAuthenticator.extra_authorize_params = {'access_type': 'offline', 'approval_prompt': 'force'}
```

**Comment by minrk** [comment]
Thanks for your patience!

**Comment by PaulMazzuca** [comment]
I think this feature is a great idea. I have needed it for a long time.  Has anyone been able to get it to work though?  It seems that when I test the login, I never get the consent screen, and therefore never get the refresh token.

After reviewing the docs https://developers.google.com/identity/protocols/oauth2/web-server, shouldn't the values be

```python
c.JupyterHub.authenticator_class.extra_authorize_params = {'access_type': 'offline', 'prompt': 'consent'}
```
Though, this does not work either.

**Comment by missingcharacter** [comment]
Hi @PaulMazzuca,

Regarding scopes, I did test it with the default scopes (`['openid', 'email']`).
Regarding `extra_authorize_params` I was not able to make it work with `{'access_type': 'offline', 'prompt': 'consent'}` and that's why I went with `{'access_type': 'offline', 'approval_prompt': 'force'}`.

**Comment by PaulMazzuca** [comment]
Interesting, perhaps there are some difference in our setups. I am using the latest OAuthenticator and JupyterHub. As I learn more about my authentication setup, I will share it.  Thanks again for this PR!

**Comment by missingcharacter** [comment]
@PaulMazzuca you can see my exact setup here: https://github.com/nxtlytics/ivy-jupyterhub/blob/master/jupyterhub_config.py#L151-L152 and container here: https://github.com/nxtlytics/ivy-base-containers/tree/master/jupyterhub-singleuser-base

**Comment by PaulMazzuca** [comment]
I don't see anything unusual. Perhaps Google is still allowing the previous keyword of "approval_prompt", though not sure why you wouldn't be able to use the new keyword "prompt".  I will keep on testing on my end to see if I can figure out what is going on.

**Comment by snickell** [comment]
After upgrading to oauthenticator 0.12, I'm getting exceptions in this code preventing my service from logging in: https://github.com/jupyterhub/oauthenticator/issues/392

**Comment by snickell** [comment]
I'm not using refresh_tokens fwiw
